src/cryptonote_core/cryptonote_core.cpp

```diff
@@ -1133,32 +1142,64 @@ namespace cryptonote
     boost::filesystem::path path(epee::string_tools::get_current_module_folder());
     path /= filename;
 
-    crypto::hash file_hash;
-    if (!tools::sha256sum(path.string(), file_hash) || (hash != epee::string_tools::pod_to_hex(file_hash)))
+    m_update_mutex.lock();
+
+    if (m_update_download != 0)
     {
-      MCDEBUG("updates", "We don't have that file already, downloading");
-      if (!tools::download(path.string(), url))
-      {
-        MCERROR("updates", "Failed to download " << url);
-        return false;
-      }
-      if (!tools::sha256sum(path.string(), file_hash))
+      m_update_mutex.unlock();
+      MCDEBUG("updates", "Already downloading update");
+      return true;
+    }
+
+    try
+    {
+      crypto::hash file_hash;
+      if (!tools::sha256sum(path.string(), file_hash) || (hash != epee::string_tools::pod_to_hex(file_hash)))
       {
-        MCERROR("updates", "Failed to hash " << path);
-        return false;
+        MCDEBUG("updates", "We don't have that file already, downloading");
+        m_last_update_length = 0;
+        m_update_download = tools::download_async(path.string(), url, [this, hash](const std::string &path, const std::string &uri, bool success) {
+          if (success)
+          {
+            crypto::hash file_hash;
+            if (!tools::sha256sum(path, file_hash))
+            {
+              MCERROR("updates", "Failed to hash " << path);
+            }
+            if (hash != epee::string_tools::pod_to_hex(file_hash))
+            {
+              MCERROR("updates", "Download from " << uri << " does not match the expected hash");
+            }
+            MCLOG_CYAN(el::Level::Info, "updates", "New version downloaded to " << path);
+          }
+          else
+          {
+            MCERROR("updates", "Failed to download " << uri);
+          }
+          m_update_download = 0;
```

Comment:
Crap, a destructor would have to lock because the callback modifies the value. Well, this stinks either way. The likelihood of stop being skipped right is now is low at least.